src/System.Private.ServiceModel/src/System/ServiceModel/Security/WSSecurityOneDotZeroSendSecurityHeader.cs

```diff
@@ -2,20 +2,819 @@
+            effectiveWriter.WriteEndElement();
+            effectiveWriter.EndCanonicalization();
+            effectiveWriter.Flush();
+
+            if (!ReferenceEquals(effectiveWriter, writer))
+            {
+                Fx.Assert(canonicalBuffer != null, "Canonical buffer cannot be null.");
+                canonicalBuffer.CloseSection();
+                canonicalBuffer.Close();
+                XmlDictionaryReader dicReader = canonicalBuffer.GetReader(0);
+                writer.WriteNode(dicReader, false);
+                dicReader.Close();
+            }
+
+            hash = hashStream.FlushHashAndGetValue();
+
+            return headerId;
+        }
+
+        private string GetSignatureStream(MessageHeader header, string headerId, IPrefixGenerator prefixGenerator, XmlDictionaryWriter writer, out Stream stream)
+        {
+            stream = new MemoryStream();
+            XmlDictionaryWriter effectiveWriter;
+            XmlBuffer canonicalBuffer = null;
+
+            if (writer.CanCanonicalize)
+            {
+                effectiveWriter = writer;
+            }
+            else
+            {
+                canonicalBuffer = new XmlBuffer(int.MaxValue);
+                effectiveWriter = canonicalBuffer.OpenSection(XmlDictionaryReaderQuotas.Max);
+            }
+
+            effectiveWriter.StartCanonicalization(stream, false, null);
+
+            header.WriteStartHeader(effectiveWriter, this.Version);
+            if (headerId == null)
+            {
+                headerId = GenerateId();
+                this.StandardsManager.IdManager.WriteIdAttribute(effectiveWriter, headerId);
+            }
+            header.WriteHeaderContents(effectiveWriter, this.Version);
+            effectiveWriter.WriteEndElement();
+            effectiveWriter.EndCanonicalization();
+            effectiveWriter.Flush();
+
+            if (!ReferenceEquals(effectiveWriter, writer))
+            {
+                Fx.Assert(canonicalBuffer != null, "Canonical buffer cannot be null.");
+                canonicalBuffer.CloseSection();
+                canonicalBuffer.Close();
+                XmlDictionaryReader dicReader = canonicalBuffer.GetReader(0);
+                writer.WriteNode(dicReader, false);
+                dicReader.Close();
+            }
+
+            stream.Position = 0;
+
+            return headerId;
+        }
+
+        private void AddSignatureReference(MessageHeader header, string headerId, IPrefixGenerator prefixGenerator, XmlDictionaryWriter writer)
+        {
+            byte[] hashValue;
+            headerId = GetSignatureHash(header, headerId, prefixGenerator, writer, out hashValue);
+            var reference = new Reference();
+            reference.DigestMethod = AlgorithmSuite.DefaultDigestAlgorithm;
+            reference.DigestValue = hashValue;
+            reference.Id = headerId;
+            _signedXml.AddReference(reference);
```

Comment:
I don't know your model, so I'll say "The other AddSignatureReference added an exc-c14n transform, should this one, too?" -- I'm having trouble aligning this with netfx to understand if there are two different types of thing.

Reply:
The difference is here we are adding a reference where the digest has already been calculated. We don't pass in the source data and have SignedXml apply the transforms and run the digest algorithm. In the other usages, we supply a memory stream which contains the XML to be signed so the transform needs to be provided. A lot of this code can probably be optimized, I was going for correctness first and then once everything is working, improve performance.

---
In reply to: [274064455](https://github.com/dotnet/wcf/pull/3538#discussion_r274064455) [](ancestors = 274064455)